### DISCUSSION TITLE
Nested embeddable of related JPA entity not initialized property

### DISCUSSION THREAD

**Discussion by odrotbohm** [querydsl/querydsl#2935] [Q&A]
Assume the following entities and embeddables:

```java
@Entity
class MyRoot {
  @OneToOne MyEntity myEntity;
}

@Entity
class MyEntity {
  MyEmbeddable embeddable;
}

@Embeddable
class MyEmbeddable {
  MyNestedEmbeddable nested;
}

@Embeddable
class MyNestedEmbeddable {

}
```

Depending on where you start in the reference chain, the metamodel property for `MyNestedEmbeddable` is null:

```java
assertNotNull(QMyEmbeddable.myEmbeddable.nested); // green
assertNotNull(QMyEntity.myEntity.embeddable.nested); // green
assertNotNull(QMyRoot.myRoot.myEntity.embeddable.nested); //red
```

Find a reproducing project here: [querydsl-jpa-embeddable.zip](https://github.com/querydsl/querydsl/files/5478105/querydsl-jpa-embeddable.zip)

**Comment by jwgmeligmeyling** [comment]
Hi @odrotbohm , I'm not sure if this is really a bug. However, its certainly annoying and error prone.

QueryDSL initializes the metamodel a single level deep. Usually this is enough, for example associations have to be joined in JPQL anyway. So instead of writing `root.myEntity.association.field` one would for example write `.innerJoin(root.myEntity.assocation, association).select(association.field)` anyway, thus working around this limitation. For nested embeddables this is less trivial.

To make sure the metamodel is initialized with `QMyRoot.myRoot.myEntity.embeddable.nested`, you can use `@QueryInits`:

```java
@Entity
class MyRoot {
  @OneToOne
  @QueryInits({"*", "*.*" /* default query inits */, "embeddable.nested"})
  MyEntity myEntity;
}
```

Hope this helps!
